ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,445 @@
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.confluent.ksql.util.QueryMetadata;
+import java.net.URI;
+import java.net.URL;
+import java.time.Clock;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.state.HostInfo;
+import org.apache.kafka.streams.state.StreamsMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The heartbeat mechanism consists of three periodic tasks running at configurable time intervals:
+ * 1. Send heartbeats: Broadcast heartbeats to remote Ksql hosts.
+ * 2. Process received heartbeats: Determine which remote host is alive or dead.
+ * 3. Cluster membership: Discover the Ksql hosts that are part of the cluster.
+ */
+@SuppressWarnings("checkstyle:ClassDataAbstractionCoupling")
+public final class HeartbeatHandler {
+
+  private final KsqlEngine engine;
+  private final ServiceContext serviceContext;
+  private final HeartbeatConfig configs;
+  private final ConcurrentHashMap<String, TreeMap<Long, HeartbeatInfo>> receivedHeartbeats;
+  private final ConcurrentHashMap<String, HostStatusEntity> hostsStatus;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final ServiceManager serviceManager;
+  private final Clock clock;
+  private final HostInfo localHostInfo;
+  private final String localHostString;
+  private final URL localURL;
+
+  private static final Logger log = LoggerFactory.getLogger(HeartbeatHandler.class);
+
+  public static HeartbeatHandler.Builder builder() {
+    return new HeartbeatHandler.Builder();
+  }
+
+  private HeartbeatHandler(final KsqlEngine engine,
+                           final ServiceContext serviceContext,
+                           final URI localServerURI,
+                           final HeartbeatConfig configs) {
+
+    this.engine = requireNonNull(engine, "engine");
+    this.serviceContext = requireNonNull(serviceContext, "serviceContext");
+    this.configs = requireNonNull(configs, "configuration parameters");
+    requireNonNull(localServerURI, "localServerURI");
+    this.scheduledExecutorService = Executors.newScheduledThreadPool(configs.threadPoolSize);
+    this.serviceManager = new ServiceManager(Arrays.asList(
+        new DiscoverClusterService(), new SendHeartbeatService(), new CheckHeartbeatService()));
+    this.receivedHeartbeats = new ConcurrentHashMap<>();
+    this.hostsStatus = new ConcurrentHashMap<>();
+    this.clock = Clock.systemUTC();
```

Comment:
The javadocs suggest using this method instead of depending on the hard-coded timezone of the application. Since all times will be converted to UTC, what is the problem? Also, this addresses the issue of not all servers being in the same timezone which if they are not , makes the heartbeat processing erroneous.

Reply:
Tracked in issue: #4240